File: client_golden_light/customizations_for_golden_light/report/cash_accounts_balance/cash_accounts_balance.js
// Copyright (c) 2022, Peniel Technology LLC and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Cash Accounts Balance"] = {
	"filters": [

		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"reqd": 1,
			"default": frappe.defaults.get_user_default("Company")
		},

		{
			"fieldname": "account",
			"label": __("Account"),
			"fieldtype": "Link",
			"options": "Account",
			"reqd": 1,
			"default": '1100-1300 - Cash and Cash Equivalents - GL'
		},
		{
			"fieldname": "from_date",
			"fieldtype": "Date",
			"default": frappe.datetime.year_start(),
			"hidden": 1,
		},
		{
			"fieldname": "to_date",
			"label": __("Balance Date"),
			"fieldtype": "Date",
			"reqd": 1,
			"default": frappe.datetime.get_today(),
		},
		{
			fieldname: "include_suppliers",
			label: __("Include Suppliers"),
			fieldtype: "MultiSelectList",
			get_data(data) {
				return frappe.db.get_link_options("Supplier", data);
			},
		}
	],
	onload(report) {
		report.set_filter_value("include_suppliers", ["مكتب رمضاني للصرافة", "شركة الصراف للصرافة"]);
	},
	formatter: function(value, row, column, data, default_formatter) {
		if (data && column.fieldname=="account") {
			column.link_onclick = "open_general_ledger(" + JSON.stringify(data) + ")";
		}

		value = default_formatter(value, row, column, data);
		return value;
	},
};


function open_general_ledger(data) {
	if (!data.account) return;

	if(data["party_type"]) {
		frappe.route_options = {
			party_type: data["party_type"],
			party: data.account,
			from_date: frappe.datetime.year_start(),
			to_date: frappe.query_report.get_filter_value("to_date"),
		};
	} else {
		frappe.route_options = {
			account: data.account,
			from_date: frappe.datetime.year_start(),
			to_date: frappe.query_report.get_filter_value("to_date"),
		};
	}
	frappe.set_route("query-report", "General Ledger");
}

File: client_golden_light/customizations_for_golden_light/report/cash_accounts_balance/cash_accounts_balance.py
# Copyright (c) 2022, Peniel Technology and contributors
# For license information, please see license.txt

# import frappe
import frappe
from erpnext.accounts.utils import get_balance_on, get_children
from frappe import _


def execute(filters=None):
    columns, data = [], []

    columns = get_columns()

    cash_accounts = get_sub_accounts(filters.get("account"), list=[])

    for account in cash_accounts:
        balance = get_balance_on(account, filters.get("to_date"), in_account_currency=False)
        account_type = frappe.db.get_value("Account", account, "account_type")
        data.append(
            {
                "account": account,
                "balance_amount": balance,
                "account_type":account_type
            }
        )

    get_suppliers(data, filters)

    return columns, data


def get_conditions(filters):
    pass


def get_sub_accounts(account, list):
    accounts = frappe.db.get_list(
        "Account", {"parent_account": account}, ["name", "is_group"], order_by="name"
    )
    for row in accounts:
        if row.is_group == 1:
            get_sub_accounts(row.name, list)
        else:
            list.append(row.name)
    return list


def get_columns():
    """Creates a list of dictionaries that are used to generate column headers of the data table."""
    return [
        {
            "fieldname": "account",
            "label": _("Account"),
            "fieldtype": "Link",
            "options": "Account",
            "width": 350,
        },
        {
            "fieldname": "balance_amount",
            "label": _("Balance Amount"),
            "fieldtype": "Currency",
            "width": 150,
        },
        {
            "fieldname": "account_type",
            "label": _("Account Type"),
            "fieldtype": "Data",
            "width": 150,
        },
    ]


def get_suppliers(data, filters=None):
    if not (filters and filters.get("include_suppliers")):
        return

    suppliers = filters.get("include_suppliers")
    account = filters.get("account")

    # if not account:
    #     frappe.throw(_("Account is missing in filters."))

    for supplier in suppliers:
        # Fetch balance amount
        balance_amount = get_balance_on(
            party_type="Supplier", party=supplier, date=filters.get("to_date")
        )
        
        # Fetch account type
        # print(account)
        account_type = frappe.db.get_value("Account", account, "account_type")

        # Debugging logs
        # frappe.log_error(f"Account: {account}, Account Type: {account_type}, Supplier: {supplier}", "Debug Info")

        # Append supplier data to the list
        data.append({
            "account": account,
            "balance_amount": balance_amount,
            "party_type": "Supplier",
            "account_type": account_type,
        })
